File: lang/java/src/206-reverse-linked-list.java
/* 206. Reverse Linked List */

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    /* Iterative solution */ 
    public ListNode reverseList(ListNode head) {
        ListNode prev = null, curr = head, next = null;
        while (curr != null) {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        head = prev;
        return head;
    }

    /* Recursive solution */
    public ListNode recursiveReverse(ListNode head) {
        ListNode curr = head;
        return helper(null, curr, null);
    }
    
    private ListNode helper(ListNode prev, ListNode curr, ListNode next) {
        if (curr == null) 
            return prev;
        
        next = curr.next;
        curr.next = prev;
        prev = curr;
        curr = next;
        
        return helper(prev, curr, next);
    }
    
    private void debug(ListNode n) {
        while (n != null) {
            System.out.println(n.val);
            n = n.next;
        }
    }
}

/*
Given the head of a singly linked list, reverse the list, and return the reversed list.

 

Example 1:

Input: head = [1,2,3,4,5]
Output: [5,4,3,2,1]

Example 2:

Input: head = [1,2]
Output: [2,1]

Example 3:

Input: head = []
Output: []
*/

/* Bonus recursive implementation */
class Solution {

File: lang/cpp/src/206-reverse-linked-list.cpp
#include <iostream>






#include <iostream>
#include "ListNode.hpp"
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
  /** Iterative */
  ListNode* reverseList(ListNode* head) {
    ListNode* prev = nullptr;
    ListNode* next = nullptr;

    while (head != nullptr)
      {
        next = head->next;
        head->next = prev;
        prev = head;
        head = next;
      }
    return prev;
  }
};

int main()
{
  Solution s1;
  std::array<int, 5> nums = {1, 2, 3, 4, 69};
  ListNode* head = createListFromArray(nums);
  printList(head);

  ListNode* r1 = s1.reverseList(head);
  printList(r1);

  return 0;
}

#include <iostream>






int main() {
class Solution {
public:
    /** Recursive */
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = nullptr;
        function<ListNode*(ListNode*, ListNode*)>
        reverseList =
        [&reverseList](ListNode* prev, ListNode* head) -> ListNode*
        {
            if (!head) return prev;
            ListNode* next = head->next;
            head->next = prev;
            return reverseList(head, next);
        };
        return reverseList(prev, head);
    }
};
return 0;
}
